Adding function information to the /render page

AlonBilman opened the pull request:
Hey Tamir!

We've added function info to the render page - it updates only when fetching from GitHub (via URL).
Currently, most of the code is in App.svelte. 
Here are some example links to test locally (port 5173):
[1](http://localhost:5173/function-graph-overview/render/?github=https://github.com/Bennahmias/Jobbot/blob/main/bot.py%23L78&colors=dark)
[2](http://localhost:5173/function-graph-overview/render/colors=dark?github=https://github.com/Bennahmias/tetris-game/blob/main/Tetris%20(2)/game.cpp%23L405)
[3](http://localhost:5173/function-graph-overview/render/?github=https://github.com/golang/go/blob/master/lib/time/mkzip.go%23L38)
[4](http://localhost:5173/function-graph-overview/render/?github=https://github.com/arrow-py/arrow/blob/master/arrow/api.py%23L106)
[5](http://localhost:5173/function-graph-overview/render/?github=https://github.com/bobbyhadz/typescript-arrow-function-generic/blob/main/src/index.ts%23L5&colors=light)
[6](http://localhost:5173/function-graph-overview/render/?github=https://github.com/glanceapp/glance/blob/main/internal/glance/cli.go%23L24&colors=light)
[7](http://localhost:5173/function-graph-overview/render/?github=https://github.com/AlonBilman/cryptocurrency_mining_linux_processes_proj/blob/main/Miner/miner.cpp%23L88)
[8](http://localhost:5173/function-graph-overview/render/?github=https://github.com/bminor/glibc/blob/master/malloc/mtrace-impl.c%23L38&colors=light)
We are waiting for your feedback and hopefully it meets your standards.

tmr232 commented on src/render/src/App.svelte:
```diff
@@ -206,6 +220,56 @@ async function createCFG(params: Params): Promise<CFG> {
+  };
+}
+
+function extractFunctionName(func: SyntaxNode, language: Language): string {
```
A few notes on this one:
- This function seems generally useful, and not just in the /render page. As such, it should be moved into the `src/control-flow` directory.
- Please add tests for all supported languages and function types. See [`functionNodeTypes`](https://github.com/tmr232/function-graph-overview/blob/181173fcae2b66a2296e85c7a7fe7dd79f5679d5/src/control-flow/cfg.ts#L56) for all supported function types.
- What happens when a function is anonymous?

tmr232 commented on src/render/src/App.svelte:
```diff
@@ -206,6 +220,56 @@ async function createCFG(params: Params): Promise<CFG> {
+    
+    default: // For any function type (C/C++/Go/Python or other non-arrow function in TypeScript/TSX)
+      return func
+        .descendantsOfType(["identifier", "field_identifier", "property_identifier"])
```
How does this work with nested functions?

tmr232 commented on src/render/src/App.svelte:
```diff
@@ -216,6 +280,9 @@ async function render() {
 
     const cfg = await createCFG(params);
+
+    updateCFGMetadata(cfg);
```
Why is this done separately from updating the function metadata?

tmr232 commented on src/render/src/App.svelte:
```diff
@@ -206,6 +220,56 @@ async function createCFG(params: Params): Promise<CFG> {
+  }
+}
+
+function updateFunctionMetadata(func: SyntaxNode, language: Language) {
```
I prefer functions that return values to functions that modify global state. Especially when modifying _part_ of an object.
I think both "update" functions need to be merged into one, and return the metadata object.

At the place where it is called, it can be used to update the global variable.

AlonBilman replied:
Alright,
the two "update" functions will be merged into a single pure function

tmr232 commented on src/render/src/App.svelte:
```diff
@@ -272,20 +339,30 @@ function makeZoomable() {
+    >Open Code</button>
     <button onclick={saveSVG}>Download SVG</button>
   </div>
+  {#if rawSVG}
```
This _works_, but hinges on an unrelated implementation detail. 
A better way to do this would be to have the metadata object as a potentially `undefined` variable, and checking that.

Also - how are we handling `graph` rendering, vs `github`?

tmr232 commented on src/render/src/App.svelte:
```diff
@@ -206,6 +220,56 @@ async function createCFG(params: Params): Promise<CFG> {
+function updateCFGMetadata(CFG: CFG) {
+  const nodeCount : number = CFG.graph.order;
+  const edgeCount : number = CFG.graph.size;
+  const cyclomaticComplexity : number = CFG.graph.size - nodeCount + 2; 
```
Please document the calculation. A link to a Wikipedia page would be good.

tmr232 commented on src/render/src/app.css:
```diff
@@ -1,4 +1,47 @@
 body {
   margin: 0;
   background: black;
+  font-family: Arial, Helvetica, sans-serif;
```
Let's a nice monospaced font instead. This _is_ a programmer's tool, after all.

tmr232 reviewed (changes requested):
Hi Alon,

Thanks for your contribution!

I went over it, and there are a few things that need more work.
There's a bit of code to change and move around, and tests that need to be written.

Additionally, I think there are visual-design changes required:
- The current color is shared for both the dark and light themes, and only _kinda_ works with either. I think it is better to use different colors for the different themes. This is especially important on the light theme, that currently has very low contrast of text and background.
- As this is a programming tool, and we're presenting data, I think a monospaced font will be a better fit.
- There is currently a lot of spacing in the display. Between the lines and between the labels and the info. I think it should be tightened.
- I think the text should be selectable. There is no reason to prevent selection.
- I think it'll be better for the data to be presented under the controls and not on the opposite side
- Maybe it's worth it to add the project and filepath as well as the function name?

![image](https://github.com/user-attachments/assets/4973cd88-2373-4bbb-bdf9-b37c16c2a015) 
![image](https://github.com/user-attachments/assets/83a183d6-cae4-4440-8d44-8d6af3bddd5d)

If you have any disagreement regarding the visual side - please let me know!

Last but not least - please make sure you run the linters (`bun lint`) and they pass before submitting your code. Note that this is critical, and in most cases I would not review code that does not pass the automated checks.

AlonBilman commented on src/render/src/app.css:
```diff
@@ -1,4 +1,47 @@
+  align-items: center;
+  justify-content: center;
+  width: 100dvw;
+  height: 100dvh;
```
Oh, alright. Thank you for the link.

AlonBilman commented on CHANGELOG.md:
```diff
@@ -13,6 +13,7 @@ Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how
   and cases after the catch-all are removed as dead code. 
 - C++ learned `co_yield` and `co_return`
 - Python learned `assert`
+- Function info and stats to the /render page. 
```
Sure thing

AlonBilman commented:
Hey Tamir!

First of all, thank you for reviewing the code!

Regarding linting: 
When running `bun lint`, it modifies the entire codebase, resulting in 100+ changes. 
Is there a way to apply the linter to a specific file? Running `bun lint <file path>` also seems to modify everything. 

Secondly, in `CONTRIBUTING.md`, there is a mention of the command `bun check` but I couldn’t get it to work. It says there is no script called "check." 
Could you please clarify how to run this command correctly? - I believe this is the reason the code didn't pass the automation.

Regarding the visual aspects - any preferences you have will be applied

Of course, all your comments will be addressed, fixed, and modified accordingly.

Thanks again for the review! changes are on the way :)

tmr232 commented:
> Regarding linting: When running `bun lint`, it modifies the entire codebase, resulting in 100+ changes. Is there a way to apply the linter to a specific file? Running `bun lint <file path>` also seems to modify everything.

The linters passing for all files are required for merging the code, you can see the failing checks at https://github.com/tmr232/function-graph-overview/actions/runs/13949594615/job/39061318568?pr=134
If you're getting a large number of changes, especially in files you did not edit yourself, it is likely that your tools are not up to date as I recently replaced prettier with biome. Try running `bun install` again and then `bun lint`.
 
> Secondly, in `CONTRIBUTING.md`, there is a mention of the command `bun check` but I couldn’t get it to work. It says there is no script called "check." Could you please clarify how to run this command correctly? - I believe this is the reason the code didn't pass the automation.

That's my bad - I changed the command from `bun check` to `bun lint` and forgot to update the docs.

tmr232 commented on src/test/functionDetails.test.ts:
```diff
@@ -0,0 +1,275 @@
+  const func = funcIterator.next().value;
+  expect(extractFunctionName(func, "Go")).toBe("Anonymous");
+});
+test("Go: nested function", () => {
+  const code = `func outer() {
+    func inner() {
+      fmt.Println("Inner function")
+    }
+    inner()
+  }`;
+  const funcIterator = iterFunctions(code, "Go");
+  const func = funcIterator.next().value;
+  expect(extractFunctionName(func, "Go")).toBe("outer");
+});
```
This seems to be testing the order of function iteration, not the name extraction mechanism.

tmr232 commented on src/test/functionDetails.test.ts:
```diff
@@ -0,0 +1,275 @@
+  expect(extractFunctionName(func, "C++")).toBe("add");
+});
+// Python Tests
+test("Python: function_definition", () => {
+  const code = `def now(tz: Optional[TZ_EXPR] = None) -> Arrow:
+    """Calls the default :class:ArrowFactory <arrow.factory.ArrowFactory> now method."""
+    return _factory.now(tz)`;
+  const funcIterator = iterFunctions(code, "Python");
+  const func = funcIterator.next().value;
+  expect(extractFunctionName(func, "Python")).toBe("now");
+});
+test("Python: function_definition", () => {
```
Please avoid giving multiple tests the same description.
If they are testing the same thing it's pointless duplication.
If they are testing different things then the description is not good enough.

tmr232 commented on src/control-flow/function-utils.ts:
```diff
@@ -0,0 +1,249 @@
+/**
+ * Extracts the name of a node by searching for a child node with a specific type.
+ *
+ * @param {SyntaxNode} func - The syntax node to search within.
+ * @param {string} type - The type of the child node to extract the name from.
+ * @returns {string | undefined} The name of the node, or `undefined` if not found.
```
The `{type}` part is not needed for TypeScript code.

tmr232 commented on src/control-flow/function-utils.ts:
```diff
@@ -0,0 +1,249 @@
+  type: string,
+): string | undefined {
+  return (
+    func.namedChildren.find((child) => child && child.type === type)?.text ||
```
You can simplify with `child?.type === type`

tmr232 replied:
See `treeSitterNoNullNodes(func.namedChildren)`

AlonBilman commented:
Hey Tamir,

Thanks again for checking out the code.
There’s definitely room for improvement and we want to make sure it lines up with what you’re looking for.
The comments are helpful, but we’d benefit from a bit of extra guidance.
Would it be possible to schedule a Discord or Google Meet call? I believe it would help a lot :) 

Thanks again , happy holiday!

Bennahmias commented:
Hey Tamir,

I saw that you changed the test framework from bun:test to vitest, and updated only the src/parser-loader/bun.ts file — but not src/parser-loader/vite.ts.

I wanted to ask if you're planning to update vite.ts as well, because in src/render/src/App.svelte we’re importing initParsers and iterFunctions from file-parsing/vite.ts.

In our tests, we also import from file-parsing/vite.ts, and it’s currently failing because of the wasm path issue. When we import from file-parsing/bun.ts, it works.

Just wanted to check if you're planning to handle this too. Thanks!

tmr232 commented:
I guess the recent changes make the naming less clear...
`vite.js` is for the built files (the extension and the web version), `bun.js` is for scripts and tests.
I might merge those into a single import at some point, but don't know when that'll be.

AlonBilman commented:
Hi Tamir,

Here’s what we’ve done so far:

First, we added support for displaying both the CFG metadata and the function name in GitHub, as well as the CFG metadata in the Graph.

We used Tree-sitter queries to improve the extraction process (thanks for the suggestion).

The code for extracting function names is spread across `src/control-flow/, where each cfg-<language>.ts` file handles its own logic.

We also kept `function-utils`, which handles shared logic and dispatching based on language.

About conflicts in function names:
When Tree-sitter queries return multiple matches, we treat it as a conflict and return `<unsupported>`. This is handled by `extractTaggedValueFromTreeSitterQuery`.
(These cases are hard to predict and usually don’t result in a valid function name.)

We added 35 tests for all the scenarios we came up with, covering different languages and function types.

On the frontend side:
In `App.svelte`, we updated the `updateMetadata `function to support Graph as well, and now we're extracting the metadata based on the render type.

(Just a note: there’s currently nothing in the README that explains how to use Graph on the render page)

When the type is GitHub, we’re fetching the data twice...
Obviously, that’s not ideal. How would you handle it better?

A GIF that summarizes the changes: 
![render-vid_](https://github.com/user-attachments/assets/17013a31-0f3d-4aee-bff5-86d2933d823e)

AlonBilman commented:
Also, when I commit my changes, should I also include the snapshot updates that were generated by the tests?

tmr232 commented:
> Also, when I commit my changes, should I also include the snapshot updates that were generated by the tests?

Are you seeing any changes beyond line-ending changes in the snapshots?

AlonBilman commented:
> > Also, when I commit my changes, should I also include the snapshot updates that were generated by the tests?
> 
> Are you seeing any changes beyond line-ending changes in the snapshots?

Oh, this is weird. Earlier this week, I saw a lot of changes in the snapshot files, but now, when I run the tests, it only changes the end-of-line characters, as you mentioned.
It wasn't like that - you can see it in [this commit](https://github.com/AlonBilman/function-graph-overview/commit/d145111b43e28c26840b49cbcc1225fcd68340f3), where I deleted the changes manually.

But for now, I guess it's alright.

tmr232 commented on src/control-flow/cfg-go.ts:
```diff
@@ -419,3 +423,75 @@ function processSwitchlike(
+  tag: "var.name",
+};
+
+const assignmentQueryAndTag = {
```
All three queries can be unified with an alternation in the query.

tmr232 commented on src/control-flow/cfg-go.ts:
```diff
@@ -419,3 +423,75 @@ function processSwitchlike(
+      return extractNameByNodeType(func, nodeType.fieldIdentifier);
+    case nodeType.funcLiteral:
+      // Check if the func_literal is assigned to a variable or is a standalone function
+      return (
```
It is important to limit the depth of a query, so that they properly ignore nesting.
Consider:

```go
func main() {
	var x = func() {
		y := func() {}
		y()
	}
	x()
}
```

See `maxStartDepth` (used in `block-matcher.ts`)

AlonBilman replied:
You're right. I'll check it out and make the changes.

tmr232 commented on src/control-flow/cfg-typescript.ts:
```diff
@@ -319,3 +323,82 @@ function processTryStatement(trySyntax: SyntaxNode, ctx: Context): BasicBlock {
+
+const variableDeclaratorQueryAndTag = {
+  query: `
+    (lexical_declaration
```
What of `var` definitions, and of definitions without either `var`, `let` or `const`?

tmr232 commented on src/control-flow/function-utils.ts:
```diff
@@ -0,0 +1,83 @@
+ *
+ * @param func - The syntax node to search within.
+ * @param type - The type of the child node to extract the name from.
+ * used among all languages (mostly the easy cases of extracting the name).
```
This line currently renders as part of the parameter definition, which seems odd to me.

tmr232 commented on src/control-flow/function-utils.ts:
```diff
@@ -0,0 +1,83 @@
+    .map((c) => c.node.text);
+
+  if (names.length > 1) {
+    return "<unsupported>";
```
What are we actually checking here? What is not supported, and why are we treating it differently than other cases where we failed to get a name?

AlonBilman replied:
As we discussed before, when I have something like `y, x := func() {}, func() {}`, the query for extracting the variables returns two results:
`  (short_var_declaration
      left: (expression_list
        (identifier) @name))`
It's unclear what the "name" of the function is, because there are two expressions. Since I get two matches for one query, I treat it as a conflict.

How can I handle this better?

tmr232 replied:
First thing, adding a comment that says this is helpful.

Second thing - please be consistent. There are more cases where we can't get the name of the function. Why are some just unnamed, and this is `<unsupported>`?

AlonBilman replied:
You're right , I should have included an explanation in the code itself.
Regarding the second thing:
I understand what you're saying. I thought it was different because the extraction succeeded, it just returned more than one name.
In most other cases, I only got a single name, so I could tell if the name extraction had failed.
I could return undefined instead

AlonBilman replied:
Should I just return undefined in this case and treat it as an unnamed function?
I think that makes sense

tmr232 replied:
Yep, `undefined` if we can't get the name, the name if we can, `anonymous` if there isn't one.

AlonBilman commented on src/control-flow/cfg-c.ts:
```diff
@@ -166,3 +166,17 @@ function processSwitchlike(switchSyntax: SyntaxNode, ctx: Context): BasicBlock {
   return blockHandler.update({ entry: headNode, exit: mergeNode });
 }
+
+const nodeType = {
```
I'll modify it

AlonBilman commented on src/control-flow/cfg-cpp.ts:
```diff
@@ -148,3 +149,72 @@ function processTryStatement(trySyntax: SyntaxNode, ctx: Context): BasicBlock {
+    if (operatorNode) return operatorNode.text;
+
+    // else, look for the destructor name, which is a special case because identifier returns without the ~
+    const destructorNode = func.descendantsOfType(nodeType.destructorName)[0];
```
Oh..good catch. That's because I gave destructors higher priority when extracting the function name. I'll change it, maybe it's time to try queries for the C and C++ cases.

AlonBilman commented:
Hi , 

So, you’ve added a couple of tests and I’d like to talk about one of them
Let’s look at this code you've given:


```
func main() {
	var x = func() {
		y := func() {}
		y()
	}
	x()
}
```


If you ask me what the name of this function is, I’d probably say "main"
Now, when a user gives the render page a GitHub link and a line number that points to a function, they might give the line for func main() { (let’s call that line 1).
If they want to look at the inner function, they would give the line for that nested function instead.
When we fetch the function from GitHub, we only get the function itself.  So if we want main, we get everything inside it.
But if we ask for line 2, we only get:
```
var x = func() {
	y := func() {}
        y()
}
```
Since we always want to get the outer function name for any link and line number everything works. we would get x.
If we were pointing at `y := func() {}` , we’d get y.

To be completely honest, I didn’t really understand the test. 
Why do we want to iterate inside the node and get the names of nested functions? I think (maybe) it doesn't match the render page feature?
Am I missing something here? 

(The test is in Go section)

AlonBilman commented on src/control-flow/cfg-cpp.ts:
```diff
@@ -148,3 +149,72 @@ function processTryStatement(trySyntax: SyntaxNode, ctx: Context): BasicBlock {
+  if (func.type === nodeType.lambdaExpression) {
+    // if the lambda is assigned to a variable, return the variable name
+    // otherwise return "<Anonymous>"
+    return (
```
Just wrote a comment about this, same here. 
Since each function stands alone, we get:
 ```
void f() {
    auto my_func = [](){
        [](){}();
    };
}
```
we get `f`. 
```
   auto my_func = [](){
        [](){}();
    };
```
we get ` my_func`

` [](){}();` 
we get  `<Anonymous>`

tmr232 replied:
But functions do not stand alone.
When we analyze code, we get the AST for the entire file, and get the relevant `Node` for the function.
That `Node` exists in the AST. As such - we should be able to _correctly_ query nested function and non-nested functions alike.

AlonBilman replied:
> But functions do not stand alone.
When we analyze code, we get the AST for the entire file, and get the relevant Node for the function.

For the render page, I believe we only get the function code and not the whole file (correct me if I’m wrong).
That’s why I was thinking this way
> get the relevant Node for the function.

If that were the case, the node would stand alone so querying the nested function should work exactly as I already did.
Does that make sense?
I’ll try to figure out why that’s not happening..

tmr232 commented:
> If you ask me what the name of this function is, I’d probably say "main"

That may be true, but what the test is asking is "what are the names of _all_ the functions here".

If we have a function that extracts function names, there is no reason for it to not work on nested functions. 

The test is there because in that code, I expect to be able to query the names of _all_ the functions, and get correct results (or `undefined` if we can't get that).
But instead, we get `main` for main, `y` for `x` and `y` for `y`.

Additionally, the test does not have to match the `/render` page. It has to test the function, which should work in any relevant use-case.

AlonBilman commented:
> The test is there because in that code, I expect to be able to query the names of all the functions, and get correct results (or undefined if we can't get that).
But instead, we get main for main, y for x and y for y.

This is kind of a new feature, extracting function names from within a function.
Ok, it still seems a bit odd to me that it doesn't work when we're iterating over the subnodes (right?).
There’s probably a difference here.

Let’s say I have a solution that returns a list of names:
For the render page - yeah, the first one is fine, Id always return the outer name.
The rest would be listed in the order they were declared inside the function?
I guess that makes sense.

I'll keep in touch

tmr232 commented on src/control-flow/function-utils.ts:
```diff
@@ -0,0 +1,83 @@
+  const language = func.tree.language;
+  const queryObj = new Query(language, query);
+
+  const rootNode = func.tree.rootNode;
```
I somehow missed this... You're getting the root of the file - this will never work.

Try the following:
- https://github.com/tmr232/function-graph-overview/blob/main/src/control-flow/cfg-python.ts#L359

If you print all the names you get with this, you'll see what I mean.

AlonBilman commented:
Hey Tamir,
When we started working again on function name extraction, we realized we need to change our approach so we can cover all the different cases and do it in the most generic way possible. That’s why we focused first on TypeScript - since it’s the more complex language when it comes to function types. We wanted to show you what we’ve done so far in this language, so that after you confirm it looks good, we can move on to the other languages.
In short, when the function is written in a “regular” way (meaning not assigned to a variable), it’s easy to grab the function name directly from the AST. But when functions start getting assigned to variables, things get messy. In those cases, we go up to the parent node and run different checks.
Let’s look at a few examples:

```
test("chained assignment in initializer", () => {
 const code = "let x = y = z = () => {};"; expect(namesFrom(code)).toEqual([“z”]); 
});

```

```
test("bare assignments (no declaration)", () => { 
const code = "x = () => {}, y = () => {};";
expect(namesFrom(code)).toEqual([“x”,“y”]); 
});

```
In test 1 we decided the name should be "z", because if we wanted to extract all the names it would already get very complicated and not generic. We would have to climb the tree to the ancestors and it doesn't work properly .

Another thing is that in this [example](https://tmr232.github.io/function-graph-overview/?compressed=N4IgbgpgTgzglgewHYgFwEYA0IA2BDJAcwFc9CI0QAXATwAcIYBjKOOqkbJhAEwtRAAPAAQBeYQAoAlGIB8w4AHoAVEgRVhfKhCbaewgBbQIyxQF9MwmmMkzR8gDpJhLhU9cfh3JDA0jxAEwA3O6eLnAAZhKCADwBUqFhHlAQVMRQzughzkkpaRnCwYnCZtmKigDKCMJURsJIeAC2EMJwMMIOIDSdAHScIBHIVBVwAF78IOgQjf3wjXQ4kTRoVFDEENgR+MMA7nBUTAYraxsgBnCEBouXHKir61wIOAiwaADaoA3NlGp8PXwRPDEHAcbBGQSUADEAHYAAxwuEgCyfJoTX4QHoQJCrZZgiAQgSQgAsAA4AIIAIQAzLCkZgUd8BOjMYJ9v1wVDKTSaXSGWjeBjalAEDt2fioQBWACcsKYst5IC+-L+NDgEBwPDFBJAkNheCleCYeAVSp+Ap62igjTgDW0Wqh0ICEoAbExnSbUWa-gAjF58KD2wmw4Mhj2MxXm86Xa4GUFncWEnhJ5NhiYQHjkHopEj4AN47WQp1SgJ4dCpyjpzPeGAQACO6yQTAo+ahRKbVLb5YElYxeBB0FtcEggZ1EQiVIgE67ICYOGIvmgPV9UH9I8hNICEE309n88tPT2tTX0J4wZPO7nC6gFqgNAAwgh5jhxS3CR229CiRe94ucWv0N6EoRAB35Xj0EQ2n2OC4vGBZSvBUroCSoH7viTbsGu0qyvKyKKp6AiEFAeB0AYS6GgA1oRCDEEgmqvjqUyMRQZgALpmEAA), we have two functions with two different variable names.
I noticed that in the demo, we're capturing the inner function.
Is that because that’s the one iterFunctions is catching?

In this case, it returns the name y,
because that’s the function that gets drawn - right?

The code as of right now includes a lot of tests for typescript and this is in “extractNameTypescript.test.ts”.
so for now you can ignore the other test files, and actually all the other fetch files..

The changes : 
https://github.com/AlonBilman/function-graph-overview/blob/65eb0a0677f09e6b92484d86f97a6d94feec89a1/src/control-flow/cfg-typescript.ts#L324-L472

https://github.com/AlonBilman/function-graph-overview/blob/issue%2394/src/test/extractNameTypescript.test.ts 

https://github.com/AlonBilman/function-graph-overview/blob/65eb0a0677f09e6b92484d86f97a6d94feec89a1/src/control-flow/function-utils.ts#L27-L84

tmr232 commented on src/test/extractNameTypescript.test.ts:
```diff
@@ -0,0 +1,384 @@
+    expect(extractFunctionName(func, "TypeScript")).toBe("returnInArray");
+  });
+
+  test("arrow variations", () => {
+    const code1 = "<T,>(value: T): T[] => {};";
+    expect(namesFrom(code1)).toEqual(["<anonymous>"]);
+
+    const code2 = `
+      function f() {
+        const b = n => n + 1;
+        const c = async () => { return 1; };
+        const d = () => () => {};
+      }
+    `;
+    expect(namesFrom(code2)).toEqual(["f", "b", "c", "d", "<anonymous>"]);
```
Please split this into 2 tests.

tmr232 commented on src/test/extractNameTypescript.test.ts:
```diff
@@ -0,0 +1,384 @@
+});
+
+describe("function declarations and expressions", () => {
+  test("function expression variants", () => {
+    const code1 = "const myFunction = function(name: string): string {};";
+    expect(namesFrom(code1)).toEqual(["myFunction"]);
+
+    const code2 = "const sum = function add(): number {};";
+    expect(namesFrom(code2)).toEqual(["add"]);
+
+    const code3 = "function(name: string): string {};";
+    expect(namesFrom(code3)).toEqual(["<anonymous>"]);
+  });
```
These checks repeat the same pattern. Please make them into a parameterized test instead.
Repeats.

AlonBilman replied:
No problem, will do

tmr232 commented on src/test/extractNameTypescript.test.ts:
```diff
@@ -0,0 +1,384 @@
+  test("generator and async functions", () => {
+    const code1 =
+      "const fn = function* myGenerator<T>(input: T): Generator<number> {};";
+    expect(namesFrom(code1)).toEqual(["myGenerator"]);
```
Honestly not sure what the name should be here. How did you decide?

tmr232 commented on src/control-flow/function-utils.ts:
```diff
@@ -0,0 +1,45 @@
+export function extractFunctionName(
+  func: SyntaxNode,
+  language: Language,
+): string | undefined {
```
When do we actually return `undefined` here?
I looked over all the typescript tests and didn't see it tested for.
What is the difference between `<anonymous>` and `undefined`?

AlonBilman replied:
I can’t be sure what kind of `SyntaxNode `the function will be called with, it may contain no functions at all.
In that case, it will return undefined.
Should I write a test for this, or modify the code?

Also, there are functions that aren’t supported or captured by `iterFunction`. I came across a couple of them (I’ll find an example and add it).
I figure that’s fine, since the render/demo doesn’t show them.

`<anonymous>` will be returned only for specific functions that do not have assignments.

tmr232 replied:
Given that this function is used to get the names of functions, I don't see a reason to differentiate the two cases.
If it is called with something that's not a function, we're good returning `undefined` as it is not a function. Great.
If we have a function, but can't get the name - `undefined` is also good, and the display code can decide what to do with that.

AlonBilman replied:
If I understand correctly, do you want `extractFunctionName `to not return <anonymous> at all, and instead let the display code (in `app.svelte`?) decide if a function is anonymous? 

sure, I'll do that

AlonBilman replied:
Something like this?

https://github.com/AlonBilman/function-graph-overview/blob/96f7cf9000aca3a4e1be7c3acaf1e95d001709ae/src/render/src/App.svelte#L76-L94

https://github.com/AlonBilman/function-graph-overview/blob/96f7cf9000aca3a4e1be7c3acaf1e95d001709ae/src/render/src/App.svelte#L307-L319

tmr232 commented on src/control-flow/cfg-python.ts:
```diff
@@ -624,3 +625,21 @@ function processWhileStatement(
+    functionQuery.functionDefinition,
+    functionQuery.tag,
+  );
+  return name.length > 1 ? undefined : name[0];
```
Can the `>1` case actually happen? What does it mean?

AlonBilman replied:
I've never seen it actually happen in Python or C with the queries I've made.
It was just in case something went wrong with the query and captured something I didn't want.

I'll remove it.

AlonBilman commented on src/control-flow/function-utils.ts:
```diff
@@ -0,0 +1,45 @@
+/**
+ * Extracts the name of a function based on its syntax node and language.
+ *
+ * Supports all languages defined in the Language type:
+ * - TypeScript/TSX
+ * - C
+ * - C++
+ * - Python
+ * - Go
```
sure

AlonBilman commented on src/control-flow/query-utils.ts:
```diff
@@ -0,0 +1,23 @@
+import { Query, type Node as SyntaxNode } from "web-tree-sitter";
+
+/**
+ * Extract a single tagged value from a syntax tree using a Tree-sitter query.
```
My bad, I missed it among all the modifications I've made along the way

AlonBilman commented on src/control-flow/query-utils.ts:
```diff
@@ -0,0 +1,23 @@
+ * @param tag - The capture tag name to filter by.
+ *
+ */
+export function extractTaggedValueFromTreeSitterQuery(
```
I did this in another branch earlier, it helps me a bit.
Do you want me to keep it, or better not change this file?
[Link to diffs ](https://github.com/AlonBilman/function-graph-overview/commit/ded9a5fac16d3683680e08144590400c92296830#diff-805f38fe5c940c6bfdc3ff3224db6bcd508005655bfa5bc0ea1523440c35445e ) (only `block-matcher.ts`)

tmr232 commented on src/control-flow/cfg-go.ts:
```diff
@@ -419,3 +420,112 @@ function processSwitchlike(
+  };
+
+  // := short var declaration
+  if (parent.parent?.type === "short_var_declaration") {
```
nit: Why not a switch-case here?

AlonBilman replied:
on it

tmr232 commented on src/control-flow/cfg-c.ts:
```diff
@@ -166,3 +167,39 @@ function processSwitchlike(switchSyntax: SyntaxNode, ctx: Context): BasicBlock {
+    end,
+  );
+
+  for (const node of nodes) {
+    const decl = node?.childForFieldName("declarator");
+    if (decl && decl.type === "identifier") return node;
+  }
+
+  return null;
```
Consider https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find

tmr232 commented on src/render/src/App.svelte:
```diff
@@ -106,8 +157,10 @@ async function getFunctionByLine(
 function setBackgroundColor(colors: "light" | "dark") {
```
Probably better to rename to `applyColors` or similar, now that we're not only setting the background color!

tmr232 commented on src/render/src/App.svelte:
```diff
@@ -210,16 +282,50 @@ async function createCFG(params: Params): Promise<CFG> {
+
     if (params.type === "GitHub") {
       codeUrl = params.codeUrl;
+      const { func, language } = await fetchFunctionAndLanguage(params);
```
I don't like that we're fetchign this twice, but I guess we'll have to do for now...

tmr232 commented on src/render/src/App.svelte:
```diff
@@ -289,7 +395,33 @@ onMount(() => {
     >
     <button onclick={saveSVG}>Download SVG</button>
   </div>
+  {#if functionAndCFGMetadata}
+   <!-- Metadata display -->
+{#if metadataFields.some(field => showMetadata[field.key] && field.value() !== undefined)}
+<div class="metadata" class:panel-open={isPanelOpen}>
+    {#each metadataFields.filter(field => field.value() !== undefined) as { key, label , value}}
+    {#if showMetadata[key]}
+      <span>{label}: {value()}</span>
+    {/if}
+  {/each}
+</div>
+{/if}
+    <button class="panel-toggle" onclick={togglePanel}>
+      {isPanelOpen ? '→' : '←'}
+    </button>
+    <!-- Control panel -->
+    <div class="control-panel" class:open={isPanelOpen}>
+      <h3>Display Options</h3>
+        {#each metadataFields.filter(field => field.value() !== undefined) as { key, label }}
+      <label>
+          <input type="checkbox" bind:checked={showMetadata[key]} />
+          {label}
+        </label>
+      {/each}
+    </div>
+  {/if}
```
Please fix the indentation here.

tmr232 commented on src/render/src/App.svelte:
```diff
@@ -210,16 +282,50 @@ async function createCFG(params: Params): Promise<CFG> {
+  // Update CFG metadata
+  const nodeCount: number = CFG.graph.order;
+  const edgeCount: number = CFG.graph.size;
+  const cyclomaticComplexity: number = CFG.graph.size - nodeCount + 2; // (https://en.wikipedia.org/wiki/Cyclomatic_complexity)
```
Move the comment to above the line, no need for the line to be this long...

tmr232 commented on src/render/src/App.svelte:
```diff
@@ -210,16 +282,50 @@ async function createCFG(params: Params): Promise<CFG> {
+  const cyclomaticComplexity: number = CFG.graph.size - nodeCount + 2; // (https://en.wikipedia.org/wiki/Cyclomatic_complexity)
+
+  return {
+    functionData: name && lineCount ? { name, lineCount, language } : undefined,
```
Please make this a proper `if`, or _at least_ use parenthesis around the condition. Readability counts!

tmr232 commented on src/test/extractNameGo.test.ts:
```diff
@@ -0,0 +1,159 @@
+  });
+});
+
+describe("Go: keyed elements - not supported", () => {
+  test("simple keys", () => {
+    const code = `
+      var _ = map[interface{}]func(){
+        "simple":        func() {},  
+        'r':             func() {},  
+        fmt.Sprint(1):   func() {},         
+        T(1):            func() {},        
+        [2]int{1,2}:     func() {},   
+        1 + 2:           func() {},     
+      }
+    `;
+    expect(namesFrom(code)).toEqual([
+      undefined,
+      undefined,
+      undefined,
+      undefined,
+      undefined,
+      undefined,
+    ]);
+  });
+});
```
Yes.

AlonBilman commented on src/control-flow/cfg-typescript.ts:
```diff
@@ -319,3 +320,132 @@ function processTryStatement(trySyntax: SyntaxNode, ctx: Context): BasicBlock {
+  }
+}
+
+function findVariableBinding(func: SyntaxNode): string | undefined {
```
on it
